From: David Milburn <dmilburn@redhat.com>
Date: Tue, 9 Mar 2010 17:25:21 -0500
Subject: [scsi] fixup size on read capacity failure
Message-id: <20100309172521.GA5930@dhcp-210.hsv.redhat.com>
Patchwork-id: 23528
O-Subject: [RHEL5.6 PATCH] scsi: fixup size on read capacity failure
Bugzilla: 569654
RH-Acked-by: John Feeney <jfeeney@redhat.com>
RH-Acked-by: Jarod Wilson <jarod@redhat.com>

On read capacity failure, sd_read_capacity() should not set capacity
to non-zero value. It should be set to zero so get_capacity() returns
zero in rescan_partitions(). Fujitsu reports hangs during boot while
checking partitions on bad drive, they have verified this patch
fixes the issue.

Upstream

commit 69bdd88ca2670c321fef774e77059516f836c6f2
Author: Hannes Reinecke <hare@suse.de>
Date:   Fri Sep 1 15:50:23 2006 +0200

    [SCSI] Wrong size information for devices with disabled read access

    When accessing a device with disabled read access the capacity is set
    randomly to 1GB. This makes it impossible to userspace tools to detect
    invalid device capacities.

This resolves BZ 571493, please reviw and ACK.

Thanks,
David

 drivers/scsi/sd.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/sd.c b/drivers/scsi/sd.c
index 4fd6e68..9252987 100644
--- a/drivers/scsi/sd.c
+++ b/drivers/scsi/sd.c
@@ -1258,7 +1258,7 @@ repeat:
 		/* Either no media are present but the drive didn't tell us,
 		   or they are present but the read capacity command fails */
 		/* sdkp->media_present = 0; -- not always correct */
-		sdkp->capacity = 0x200000; /* 1 GB - random */
+		sdkp->capacity = 0; /* unknown mapped to zero - as usual */
 
 		return;
 	} else if (the_result && longrc) {
